subprocess_utils.hpp

```diff
@@ -0,0 +1,219 @@
+
+#include <string>
+#include <iostream>
+
+// unix process stuff
+#include <signal.h>
+#include <unistd.h>
+#include <sys/wait.h>
+#include <sys/prctl.h>
+#include <cstring>
+
+
+namespace subprocess {
+    
+    /**
+     * A TwoWayPipe that allows reading and writing between two processes
+     * must call initialize before being passed between processes or used
+     * */
+class TwoWayPipe {
+	private:
+		//[0] is the output end of each pipe and [1] is the input end of each pipe
+		int input_pipe_file_descriptor[2];
+		int output_pipe_file_descriptor[2];
+        std::string internalBuffer;
+        bool inStreamGood = true;
+        
+	public:
+		TwoWayPipe() = default;
+		
+		/**
+		 * initializes the TwoWayPipe the pipe can not be used until this is called
+		 * */
+		void initialize(){
+			pipe(input_pipe_file_descriptor);
+			pipe(output_pipe_file_descriptor);
+		}
+	private:
+	    /**
+	     * closes the ends that aren't used (do we need to do this?
+	     * */
+		void closeUnusedEnds() {
+			//we don't need the input end of the input pipe or the output end of the output pipe
+			close(input_pipe_file_descriptor[1]);
+			close(output_pipe_file_descriptor[0]);
+		}
+
+	public:
+	
+	    /**
+         * sets this to be the child end of the TwoWayPipe
+         * linking the input and output ends to stdin and stdout/stderr
+         * */
+		bool setAsChildEnd() {
+		    int tmp[2] = {input_pipe_file_descriptor[0],input_pipe_file_descriptor[1]};
+
+		    input_pipe_file_descriptor[0] = output_pipe_file_descriptor[0];
+			input_pipe_file_descriptor[1] = output_pipe_file_descriptor[1];
+			output_pipe_file_descriptor[0] = tmp[0];
+			output_pipe_file_descriptor[1] = tmp[1];
+		    
+			dup2(input_pipe_file_descriptor[0], STDIN_FILENO);
+			dup2(output_pipe_file_descriptor[1], STDOUT_FILENO);
+			dup2(output_pipe_file_descriptor[1], STDERR_FILENO);
+
+			closeUnusedEnds();
+		}
+        
+        /**
+         * sets this pipe to be the parent end of the TwoWayPipe
+         * */
+		bool setAsParentEnd() {
+			closeUnusedEnds();
+		}
+        
+        /**
+         * writes a string to the pipe 
+         * @param input - the string to write
+         * @return the number of bytes written
+         * */
+		size_t writeP(const std::string &input) {
+			return write(output_pipe_file_descriptor[1], input.c_str(), input.size());
+		}
+
+        /**
+         * @return true unless the last call to read either failed or reached EOF
+         * */
+        bool isGood(){
+            return inStreamGood;
+        }
+
+        /**
+    	* reads up to n bytes into the internal buffer
+    	* @param n - the max number of bytes to read in
+    	* @return the number of bytes read in, -1 in the case of an error
+    	* */
+        ssize_t readToInternalBuffer()
+        {
+            char buf[256];
+            int cnt;
+            ssize_t bytesCounted = -1;
+
+            while((bytesCounted = read(input_pipe_file_descriptor[0], buf, 256))<=0){
```

Comment:
why did we switch back to the read syscall?

Reply:
https://stackoverflow.com/questions/1589168/getline-over-a-socket
Seems to suggest some issues with getline() but if we're confident that those issues have long since dissappeared I'm happy to switch back.(It does however give us more control over the pipe)

Reply:
that's a socket tho, is that relevant for what this is going to be doing? I don't think stdin/out is ever a socket, so we can safely assume they're files, so getline should work i think

Reply:
So the implementation of getline reads a single character from the pipe at once, I'm not sure if this will have a meaningful impact on speed or even if the read method is faster, but we can test it.

Reply:
Yeah, it would be good to see the difference. I haven't thought enough about exceptional cases and stuff (you log to std::cerr at one point), so let's change it when we decide about it.

We can go the python route (based on the subprocess library) where a subprocess that exits with a non-zero status will raise an exception - this includes cases such as the executable doesn't exist (we can't test this a priori without a file system race condition), crashes, etc.

If we do, the function won't need to return a value anymore I guess? Idk, we can speak about in person next time

Reply:
if exec fails we can write the failure reason to a special failure pipe, of course then the race condition move to when we read/ write to that pipe.

Reply:
does that feel like idiomatic modern c++? that's certainly the C way of doing things, but I think for exceptional circumstances we should make it throw exceptions, otherwise you'll find people when using this library, they have to include so many is_good() checks within their code

Reply:
yes, we would throw an exception, but we need to detect that exec failed using pipes, since the exception would be in a different process.

Reply:
ah, okie, this is fair

you've fixed most of these things, so the only things that are left to do are: merge the utils stuff into the single file for now, while we wait for hooks to start working; and secondly run clang-format-6.0 on your code just so that the travis thing will pass (i'm about to merge that in now)

cheers

Reply:
what style format are we using for clang-format?

Reply:
I just went with the style used in https://github.com/souffle-lang/souffle

Reply:
yep, I'm used that one, but that's in master now